common/aicore/path_finding.cpp

```diff
@@ -95,8 +95,14 @@ struct pf_map {
   struct pf_parameter params; // Initial parameters.
 };
 
+template <typename T> struct pf_map_trait {
+  using Type = std::conditional_t<std::is_const_v<std::remove_pointer_t<T>>,
+                                  const pf_map, pf_map>;
+};
+
 // Down-cast macro.
-#define PF_MAP(pfm) ((struct pf_map *) (pfm))
+#define PF_MAP(pfm)                                                         \
+  (reinterpret_cast<typename pf_map_trait<decltype(pfm)>::Type *>(pfm))
```

Comment:
1. Thinks macros are ugly
2. Hides templates in a macro

Nice solution to the const-correctness problem! Would have skipped the struct and written `using as_pf_map_t = ...` directly, but it works.

Reply:
Thinking about it now, this would be a slightly simpler solution (if I understood the problem correctly):
```cpp
/**
 * Casts to `pf_map`
 *
 * \fixme Capitalized because this used to be a macro.
 */
pf_map *PF_MAP(void *x) { return reinterpret_cast<pf_map *>(x); }

/**
 * Casts to `pf_map`
 *
 * \fixme Capitalized because this used to be a macro.
 */
const pf_map *PF_MAP(const void *x) { return reinterpret_cast<const pf_map *>(x) }
```